[#528] remove runtime dependencies to xbase.lib

cdietrich opened the pull request:
[#528] remove some runtime dependencies to xbase.lib
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

cdietrich commented on org.eclipse.lsp4j.generator/src/main/java/org/eclipse/lsp4j/generator/JsonRpcDataProcessor.xtend:
```diff
@@ -154,10 +160,9 @@ class JsonRpcDataProcessor extends AbstractClassProcessor {
 			val accessorsUtil = new AccessorsProcessor.Util(context)
 			body = '''
-				«ToStringBuilder» b = new «ToStringBuilder»(this);
+				«MoreObjects.ToStringHelper» b = «MoreObjects».toStringHelper(this);
```
alternatively we may add a copy of ToStringBuilder to our codebase

pisv replied:
> alternatively we may add a copy of ToStringBuilder to our codebase

I would vouch for that. It would be a pity to lose the nicely formatted output provided by ToStringBuilder for deeply nested structures that are so common in LSP, both when debugging and in error logs. It seems that copying ToStringBuilder would add little if anything to maintenance costs.

szarnekow replied:
I'd do the same. Copy the ToStringBuilder as a package private class.

nixel2007 commented on org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/adapters/HoverTypeAdapter.xtend:
```diff
@@ -37,15 +38,15 @@ class HoverTypeAdapter {
 		val nextToken = in.peek()
 		if (nextToken == JsonToken.STRING) {
-			val List<Either<String, MarkedString>> value = newArrayList(Either.forLeft(in.nextString))
+			val List<Either<String, MarkedString>> value = Lists.newArrayList(Either.forLeft(in.nextString))
```
`List.of`?

cdietrich replied:
is java 9+

nixel2007 replied:
when it comes to guava, less is better)

nixel2007 replied:
> is java 9+

ah, i see. sorry then

cdietrich replied:
de-guavaize would be a second step.

cdietrich replied:
did replace it with pure java

cdietrich commented on org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/adapters/HoverTypeAdapter.xtend:
```diff
@@ -37,15 +38,17 @@ class HoverTypeAdapter {
 		val nextToken = in.peek()
 		if (nextToken == JsonToken.STRING) {
-			val List<Either<String, MarkedString>> value = newArrayList(Either.forLeft(in.nextString))
+			val List<Either<String, MarkedString>> value = new ArrayList<Either<String, MarkedString>>()
```
replaced it with a bit more verbose pure java

szarnekow replied:
If verbosity is a concern, you could remove the type arguments from the constructor invocation. They would be inferred. But I doubt it matters much.

dhuebner commented on org.eclipse.lsp4j.generator/src/main/java/org/eclipse/lsp4j/generator/JsonRpcDataProcessor.xtend:
```diff
@@ -44,7 +44,12 @@ class JsonRpcDataProcessor extends AbstractClassProcessor {
+			if (purified !== null) {
+				method.removeAnnotation(purified)
+			}
+		]
```
Is that also a part of this issue?

cdietrich replied:
yes it removes the pure annotation

jonahgraham commented:
What is the state of this PR - is this something you are maintaining on the side for those that need it or does it need to get to completion one day?

cdietrich commented:
i dont know if it is properly working as expected as where else we use xbase.lib

jonahgraham commented:
OK - me neither, and I don't know how to tell. I guess if it looks good we can release it and have a trial by fire on it. We are still an incubating project.

cdietrich commented:
Guess we have to Grep generated code where Xbase and Xtend lib still are used

cdietrich commented:
@pisv @szarnekow @jonahgraham 
i have updated this pr by
- also remove the annotation for the adapters `TypeAdapterImpl` annotation
- add copies of `ToStringBuilder` so we are compatible and dont need guava dependency
- moved generator dependency to compileOnly (so no longer maven dependency to it)
- added ArchUnit tests to make sure we no longer depend on xbase.lib, xtend.lib, guava and lsp4j.generator at runtime.

what do you think. maybe we can merge this shortly after the release
cc @rgrunber this might be of interest for you on jdt.ls side

szarnekow reviewed (approved):
I like this.

cdietrich commented:
should we also remove NoAnnotationTest or keep it?

szarnekow commented:
> should we also remove NoAnnotationTest or keep it?

I don't think it does any harm but does not add any value either.

szarnekow commented:
> > should we also remove NoAnnotationTest or keep it?
> 
> I don't think it does any harm but does not add any value either.

Read: inconclusive :)

jonahgraham commented:
> what do you think. maybe we can merge this shortly after the release

Sounds like a good plan, lets just give it a few days in case we need to respin LSP4J for 2022-03 release.

rgrunber commented:
> what do you think. maybe we can merge this shortly after the release cc @rgrunber this might be of interest for you on jdt.ls side

We plan to update to LSP4J 0.20.0 soon, and once that's done we'll probably be able to remove our dependence on xbase.lib as well.

cdietrich commented:
@jonahgraham what do you think? ready to merge or wait another week?
didnt folloe if you built the patch from master or a dedicate maintenance branch

jonahgraham commented:
@cdietrich go ahead and merge. I now have the `release_0.20` branch if I need to do any further respins for 2023-03

cdietrich commented:
this pr is merged now. so using the nightly it should be possible to consume without guava, xtext and xtend
cc @rgrunber @mickaelistria
